### PR TITLE
Core

### PR DESCRIPTION
**Pull request by chemistry-sourabh**
Pushed something so that people can start discussing and  start writing interfaces.

Issue #26 
Read some of the comments I made and tear it apart as usual.

Some of the main issues I found were the interdependence of drivers especially for diskless and storage.
The solution I came up with is that each diskless driver should provide an interface which describes what it needs from the storage side.

Similarly the DHCP, PXE and TFTP part is confusing. The template is also diskless dependent and pxe related for sure. Don't know if it is dependent on DHCP also ?

Fire away!

### CODE REVIEW

File: m2/core.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+            instance_id = self.db.provisionedInstance.insert(instance_name,
+                                                             instance_type,
+                                                             self.entityId,
+                                                             network_id)
+
+            self.multi_tenancy.setup(instance_id, network_id)
+
+            clone_id = self.storage.clone(instance_id,
+                                          parent_image_id)  # creates
```
Comment (djfinn14) [inline m2/core.py]: How does the clone name get determined, is it just a configuration option?

**Comment by xuhang57** [comment]
Didn't know M2(BMI) core is so concise ;). Beautiful!

Do we have any test script to test the core code? It would help me get familiar with the code base easier.

File: m2/core.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
+# TODO Enforce Quota Checks
+# TODO Check Authorization
+class Core:
+
+    # Wrote for probably working with middleware only
+    def __init__(self, entityId, config):
+        # Load Drivers
```
Comment (xuhang57) [inline m2/core.py]: docstring in module, class, functions needed to pass pycodestyle/pylint

File: m2/core.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        self.authorization = None
+
+        # self.db = Database(config)
+        self.db = None
+
+        self.entityId = entityId
+        self.config = config
+
+    def provision(self, instance_name, parent_image_id, mac_address, nic,
+                  # provision_engine_id, Do we really need this ?
```
Comment (xuhang57) [inline m2/core.py]: maybe now it is a good time to discuss this rather than commenting here.

File: m2/core.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            target_id = self.diskless.get_target_id(instance_id)
+            self.pxe.unregister(instance_id)
+            self.pxe.register(instance_id, dest_mac_address, target_id,
+                              dest_nic, self.config.diskless.key)
+
+        except Exception:
+            pass
+
+    def deprovision(self, instance_id):
+        """
```
Comment (xuhang57) [inline m2/core.py]: a few words on what this function does would be great. same as other functions.

File: m2/core.py (hunk 1 above)
Comment (chemistry-sourabh) [reply m2/core.py]: As far as my understanding the clone name is something internal to the driver. inter driver communication should happen mostly with ids. An id is an integer.

**Comment by chemistry-sourabh** [comment]
I will add the comments shortly and no I don't have tests have yet. I will need some mock drivers to write tests or will need some mock test suite.

File: m2/core.py (hunk 3 above)
Comment (chemistry-sourabh) [reply m2/core.py]: That is why I put the comment :)

**Comment by chemistry-sourabh** [comment]
Please review what I posted. I will add some more functions tomorrow.

**Comment by xuhang57** [comment]
would someone be able to help you on getting the mock test drivers in place? I mean, it would be nice to have this PR along with some actual tests

**Comment by chemistry-sourabh** [comment]
@xuhang57 first let me finish writing all the functions and let people agree on it. After that people can write their interfaces along with some mock drivers. After that I can add some tests in this PR or in a separate PR (Depends on you guys) and merge it.

**Comment by xuhang57** [comment]
Make sense, so what you are suggesting is 

For this PR:

- [ ] Agree on how all functions in M2 core should work.

- [ ] Write tests to test the code in this PR (#27)

- [ ] Review and Merged

For other folks who wring the interfaces

- [ ] Discuss in this PR (#27) and have an agreement

- [ ] Start writing interfaces and test along with the interfaces (including mock test drivers)

- [ ] Review and merged

And what I was saying is that, for people who are not writing the interfaces and not familiar with the previous code base (ims), like me, won't be able to give a good review, since I have no idea how to use the code committed in this PR.

In short, I didn't know your plan and let's see what other people would say.  And personally, I should be able to provide a more solid review later on. Thanks! :)

File: m2/core.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        :param instance_type: BM (Bare Metal) or VM
+        :param network_id: Id of the network that has the node
+        :param extra_params: Any extra params that may be required (Ignored)
+        :return: Provisioned Instance Id
+        """
+
+        try:
+
+            # Check whether entity has access to node and network
+            self.driver.authorization.check_entity_node_access(self.entity_id,
```
Comment (nasibehteimouri) [inline m2/core.py]: Is authorization supposed to have a method "check_entity_node_access"? I have defined "authorize_access_node" that supposes to authorizes entity's access to a specific node. So, I should change naming there? I am going to raise PRs for interfaces soon.

File: m2/core.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            # To solve this
+            # * Each diskless driver should provide another interface called
+            # Diskless Storage interface that either the storage driver or some
+            # other class can implement.
+            # * There should a Diskless Storage driver for every
+            # diskless-storage combination we want to support.
+            # * The diskless storage driver is going to be passed as an
+            # argument to diskless driver during init and the diskless driver
+            # is going to use it in mount and unmount clone calls.
+            target_id = self.driver.diskless.mount_clone(instance_id, clone_id)
```
Comment (nasibehteimouri) [inline m2/core.py]: Same comment here, In the interface "target", I have a method "add_target" that seems to do the same as mount_clone.

File: m2/core.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+            # I dont see a reason yet to write a pxe driver as PXE, DHCP and
+            # TFTP are all protocols and driver is just going to generate files
+            # that follow a template. We can hardcode a driver in the driver
+            # loading phase and pass a key (self.config.diskless.key) that will
+            # tell the driver which template to use as the template is diskless
+            # dependent. Incase we need a driver in the future should be less
+            # than 10 lines.
+            # Generates the mac address and ipxe files.
+            self.driver.pxe.register(instance_id, mac_address, target_id, nic,
```
Comment (nasibehteimouri) [inline m2/core.py]: same comment as above.

File: m2/core.py (hunk 5 above)
Comment (nasibehteimouri) [reply m2/core.py]: BTW, I like "check_entity_node_access" more than mine! and will use it.

**Comment by chemistry-sourabh** [comment]
@xuhang57 you got the idea of this PR right. The discussion in this PR is going to be long so it will be a good idea if someone can maintain some notes as we will be digging into the smallest implementation details. We can discuss this PR on Friday also.

File: m2/core.py (hunk 6 above)
Comment (chemistry-sourabh) [reply m2/core.py]: I thought we agreed to call the interface "Diskless" ?  I don't have an issue with add_target.

File: m2/core.py [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+# TODO Add Admin Level Stuff
+class Core:
+    """
+    The class that will process a REST request.
+    Currently will only work with middleware.
+    """
+
+    # Can be extended for DB Auth by doing some architecture change
+    # Authentication Driver should be used in rest.py to get entity object
+    # and passed here.
```
Comment (naved001) [inline m2/core.py]: Agreed

File: m2/core.py [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+    def __init__(self, entity, config, driver):
+        """
+        Initializes DB and other parameters
+
+        :param entity: Entity Object
+        :param config: The Config Object
+        :param driver: Object whose attributes are loaded driver objects
```
Comment (naved001) [inline m2/core.py]: why does core take the driver object here? config object should have that.

File: m2/core.py [hunk 10]
```diff
@@ -0,0 +1,653 @@
+""" Core of M2 which is the backend of the REST API """
+
+
+# TODO Enforce Quota Checks
+# TODO Check Authorization
+# TODO Add Admin Level Stuff
+class Core:
```
Comment (naved001) [inline m2/core.py]: few questions:

1. Who's actually going to make an object of this class?
2. We should first load the config file somewhere, and the drivers, and somewhere we make this init this stuff. Like how we load most of our stuff when starting the HIL [server here](https://github.com/CCI-MOC/hil/blob/master/hil.wsgi)

Comment (naved001) [reply m2/core.py]: One more question, the way we do in HIL is, we don't have a class in our core, we just have methods decorated by the rest wrappers. So why are we making a class here? 

https://github.com/CCI-MOC/hil/blob/master/hil/api.py#L45

**Comment by chemistry-sourabh** [comment]
@naved001 If I understand your suggestion right then it is to combine rest.py and core.py into a single file. I feel that may be too much of a mess. 

To answer some of your questions.
I don't think you were there for the meeting where we initially discussed the folder structure (I was physically present at MOC). We decided to have an init script that will 
1. read the config file
2. create the config object
3. load the drivers
4. start flask server by passing these parameters. (We didnt really discuss about driver objects being separate from config object, I feel that config object should just have settings)

Then rest.py is going to (This is how I visualized it)
5. do auth with auth driver if no middleware is used for each request
6. create core object with parameters
7. Do operation
8. Convert return value to JSON and return with correct Status code.

I think this will answer the questions you asked.

I will cleanup the code and remove some comments which we decided on so that people don't get overwhelmed

**Comment by naved001** [comment]
> combine rest.py and core.py into a single file. I feel that may be too much of a mess.

Yeah, I am not suggesting that. I like it separate, If I were to do it, I would just copy HIL's template of the flask server.

**Comment by chemistry-sourabh** [comment]
Guys please address comments in line 42 and line 76 also, missed them in meeting.

**Comment by chemistry-sourabh** [comment]
@naved001 saw HIL's template. It is currently doing rest+core and using model for db operations. I want rest to just do HTTP request/response stuff and core to do backend stuff.

We can create one core object in rest.py instead of multiple ones. (Although it will make connection management easier if we have multiple objects)

**Comment by naved001** [comment]
The core stuff is here https://github.com/CCI-MOC/hil/blob/master/hil/api.py not in rest.py, rest.py only does flask and auth.

File: m2/core.py [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+            self.driver.storage.clone(instance_id, parent_image_id)
+
+            # Get Clone Name and Other information needed for diskless
+            clone_info = self.driver.storage.get_clone_info(instance_id)
+
+            # Add Target To Diskless
+            self.driver.diskless.add_target(instance_id, clone_info)
+
+            # Generates the mac address and ipxe files.
+            self.driver.diskless.register(instance_id, mac_address, nic)
```
Comment (naved001) [inline m2/core.py]: @apoorvemohan feels that calling all these methods in a row like this makes exception handling really difficult. This is exactly how we do it in the current version of BMI. We ought to come up with a better strategy. We will talk about it on Friday.

File: m2/core.py [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+    """
+    The class that will process a REST request.
+    Currently will only work with middleware.
+    """
+
+    # Can be extended for DB Auth by doing some architecture change
+    # Authentication Driver should be used in rest.py to get entity object
+    # and passed here.
+    # TODO Should check if entity is in quota table before proceeding
+    def __init__(self, entity, config, driver):
```
[the remark this answers is not in the document]
Comment (xuhang57) [reply m2/core.py]: +1

**Comment by chemistry-sourabh** [comment]
@naved001 @apoorvemohan I can meet you guys this week, so that we can sort it out.
